### ISSUE TITLE
Flaky test RoundRobinConnectionPoolTest.testRoundRobin()

### ISSUE THREAD

**Issue by lachlan-roberts** [eclipse/jetty.project#6217]
**Jetty version**
10.0.x

**Description**
The test `RoundRobinConnectionPoolTest.testRoundRobin()` is failing for some PR builds.

https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/jetty-10.0.x-FixTypo/1/tests

```
[55810, 55812, 55808, 55810, 55808]
Expected: <55808>
     but: was <55812>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.eclipse.jetty.http.client.RoundRobinConnectionPoolTest.testRoundRobin(RoundRobinConnectionPoolTest.java:103)
```

**Comment by lachlan-roberts** [comment]
Failed again:
https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/PR-6226/3/tests/

**Comment by lachlan-roberts** [comment]
This test only seems to fail for the FCGI transport.
Can be reproduced easily enough locally if the test is on repeat until failure.

**Comment by lachlan-roberts** [comment]
@lorban do you want to look into this one? This one seems to fail a lot on CI.
https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/PR-6569/8/tests/

**Comment by lorban** [comment]
There seems to be a race condition between the calls to `HttpChannelOverFCGI.eof()` and `HttpChannelOverFCGI.failAllContent()`. The former should always be called first but it seems that in rare occurrences it is called second.

This has a side-effect that the `Connection: close` header is added to the response, so the connection is removed from the pool, which eventually triggers the assertion error as the connections in the pool did not generate the results that were expected from a larger number of connections.

Still investigating...

**Comment by lorban** [comment]
This race condition is intrinsic to the way FCGI currently is implemented. There needs to be adjustments such as the parser doesn't run concurrently with the dispatched servlet thread anymore.

**Comment by sbordet** [comment]
Closed in favor of #6693.
